### PR TITLE
[SPARK-11149] [SQL] Improve cache performance for primitive types

### PR DESCRIPTION
**Pull request by davies**
This PR improve the performance by:

1) Generate an Iterator that take Iterator[CachedBatch] as input, and call accessors (unroll the loop for columns), avoid the expensive Iterator.flatMap.

2) Use Unsafe.getInt/getLong/getFloat/getDouble instead of ByteBuffer.getInt/getLong/getFloat/getDouble, the later one actually read byte by byte.

3) Remove the unnecessary copy() in Coalesce(), which is not related to memory cache, found during benchmark.

The following benchmark showed that we can speedup the columnar cache of int by 2x.

```
path = '/opt/tpcds/store_sales/'
int_cols = ['ss_sold_date_sk', 'ss_sold_time_sk', 'ss_item_sk','ss_customer_sk']
df = sqlContext.read.parquet(path).select(int_cols).cache()
df.count()

t = time.time()
print df.select("*")._jdf.queryExecution().toRdd().count()
print time.time() - t
```

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeFormatter.scala [hunk 1]
```diff
@@ -44,11 +45,13 @@ private class CodeFormatter {
     } else {
       indentString
     }
+    code.append(f"${currentLine}%03d ")
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeFormatter.scala]: as discussed offline, we can add `/* ... */` to still enable pasting this into an IDE.

File: sql/core/src/main/scala/org/apache/spark/sql/columnar/ColumnType.scala [hunk 2]
```diff
@@ -28,6 +28,36 @@ import org.apache.spark.sql.types._
 import org.apache.spark.unsafe.Platform
 import org.apache.spark.unsafe.types.UTF8String
 
+
+/**
+ * A help class for fast reading Int/Long/Float/Double from ByteBuffer in native order.
+ */
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/columnar/ColumnType.scala]: put a big warning here that this only works with HeapByteBuffer.

File: sql/core/src/main/scala/org/apache/spark/sql/columnar/GenerateColumnAccessor.scala [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+ * Generates bytecode for an [[ColumnarIterator]] for columnar cache.
+ */
+object GenerateColumnAccessor extends CodeGenerator[Seq[DataType], ColumnarIterator] with Logging {
+
+  protected def canonicalize(in: Seq[DataType]): Seq[DataType] = in
+  protected def bind(in: Seq[DataType], inputSchema: Seq[Attribute]): Seq[DataType] = in
+
+  protected def create(columnTypes: Seq[DataType]): ColumnarIterator = {
+    val ctx = newCodeGenContext()
+    val (creaters, accesses) = columnTypes.zipWithIndex.map { case (dt, index) =>
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/columnar/GenerateColumnAccessor.scala]: creators

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/columnar/GenerateColumnAccessor.scala]: actually probably more clear to say

initializeAccessors and extractors

File: sql/core/src/main/scala/org/apache/spark/sql/columnar/GenerateColumnAccessor.scala [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+          this.columnTypes = columnTypes;
+          this.columnIndexes = columnIndexes;
+        }
+
+        public boolean hasNext() {
+          if (currentRow < totalRows) {
+            return true;
+          }
+          if (!input.hasNext()) {
+            return false;
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/columnar/GenerateColumnAccessor.scala]: when can this happen?

File: sql/core/src/main/scala/org/apache/spark/sql/columnar/GenerateColumnAccessor.scala [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+          }
+
+          ${classOf[CachedBatch].getName} batch = (${classOf[CachedBatch].getName}) input.next();
+          currentRow = 0;
+          totalRows = batch.count();
+          for (int i=0; i<columnIndexes.length; i++) {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/columnar/GenerateColumnAccessor.scala]: might as well fit our own style here: add space around = and <

File: sql/core/src/main/scala/org/apache/spark/sql/columnar/InMemoryColumnarTableScan.scala [hunk 6]
```diff
@@ -43,7 +42,7 @@ private[sql] object InMemoryRelation {
       tableName)()
 }
 
-private[sql] case class CachedBatch(buffers: Array[Array[Byte]], stats: InternalRow)
+private[sql] case class CachedBatch(count: Int, buffers: Array[Array[Byte]], stats: InternalRow)
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/columnar/InMemoryColumnarTableScan.scala]: let's add classdoc for count, buffers, and stats.

Without looking at the code it is less clear what count is here. (Is it the total number of rows?)

File: sql/core/src/main/scala/org/apache/spark/sql/columnar/GenerateColumnAccessor.scala (hunk 4 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/columnar/GenerateColumnAccessor.scala]: This will happen at the end of Iterator[CachedBatch]

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/columnar/GenerateColumnAccessor.scala]: As discussed offline, the confusion is coming from "totalRows". Maybe name that "numRowsInBatch"?

**Comment by rxin** [comment]
LGTM - although I didn't look super closely so might be good for an extra pair of eyes too.

**Comment by davies** [comment]
I'm going to merge this first, unblock me to work on output UnsafeRow for columnar cache, because having Unsafe format inside a MutableRow could result unexpected behavior, any new comments will be address in follow up PR.

File: sql/core/src/main/scala/org/apache/spark/sql/columnar/ColumnType.scala [hunk 7]
```diff
@@ -219,11 +251,11 @@ private[sql] object DOUBLE extends NativeColumnType(DoubleType, 8) {
...
-    buffer.getDouble()
+    ByteBufferHelper.getDouble(buffer)
   }
 
   override def extract(buffer: ByteBuffer, row: MutableRow, ordinal: Int): Unit = {
-    row.setDouble(ordinal, buffer.getDouble())
+    row.setDouble(ordinal, ByteBufferHelper.getDouble(buffer))
   }
 
   override def setField(row: MutableRow, ordinal: Int, value: Double): Unit = {
```
Comment (tedyu) [inline sql/core/src/main/scala/org/apache/spark/sql/columnar/ColumnType.scala]: Around line 332, there is call to buffer.getShort()

Is it worth adding corresponding method to ByteBufferHelper ?

If so, I can send a PR.

Thanks

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/columnar/ColumnType.scala]: I think it does not worth it.

File: sql/core/src/main/scala/org/apache/spark/sql/columnar/GenerateColumnAccessor.scala [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+        public SpecificColumnarIterator() {
+          this.nativeOrder = ByteOrder.nativeOrder();
+          this.buffers = new byte[${columnTypes.length}][];
+
+          ${initMutableStates(ctx)}
+        }
+
+        public void initialize(scala.collection.Iterator input, MutableRow mutableRow,
+                               ${classOf[DataType].getName}[] columnTypes, int[] columnIndexes) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/columnar/GenerateColumnAccessor.scala]: wrong indent here

**Comment by cloud-fan** [comment]
LGTM

**Pull request closed without merging**